subprojects/core/src/integTest/groovy/org/gradle/api/tasks/ExecIntegrationTest.groovy

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+import org.gradle.integtests.fixtures.AbstractConsoleFunctionalSpec
+import spock.lang.Issue
+
+class ExecIntegrationTest extends AbstractConsoleFunctionalSpec {
+    @Issue("https://github.com/gradle/gradle/issues/2009")
+    def "Exec task output is grouped with it's task output"() {
+        given:
+        buildFile << """apply plugin: 'java'
+
+task run(type: Exec) {
+    dependsOn 'compileJava'
+    commandLine 'java', '-cp', "\$sourceSets.main.runtimeClasspath.asPath", 'Main'
+}
+"""
+
+        and:
+        file("src/main/java/Main.java") << """public class Main {
+    public static void main(String[] args) {
+        System.out.println("Hello, World!");
+    }
+}"""
+
+        when:
+        succeeds("run")
+
+        then:
+        (result.output =~ /(?ms)(> Task :run.*?Hello, World!)/).find()
```

Comment:
You are right that we could have a generic pattern for matching only control characters so we can assert the two strings are actually following each other.

> 1. A build that has logging statements in the build script itself (System.out and logger.warn) -- there should be some way to tell that we're not mixing things.
I don't feel this will be happening in this exact test. We are only testing for logs from `exec` are attached to a task.

> 2. A build that has multiple tasks that log things in parallel. This needs to probably be an Exec-like task with another task that uses `logger/System.out`.
Flushing the logs in the console is serial even if Gradle is executing in parallel.

> 3. As I said above, a build that uses JavaExec/Exec/project.exec/project.javaExec.
I will add more coverage for those scenarios.

> What's the scope of things we think are not fixed by this change right now?
Anything that has a similar reproducing pattern, for example https://github.com/gradle/gradle/pull/2052